daiquiri/formatter.py

```diff
@@ -35,19 +37,111 @@ class ColorFormatter(logging.Formatter):
 
     COLOR_STOP = '\033[0m'
 
-    def format(self, record):
+    def add_color(self, record):
         if getattr(record, "_stream_is_a_tty", False):
             record.color = self.LEVEL_COLORS[record.levelno]
             record.color_stop = self.COLOR_STOP
         else:
             record.color = ""
             record.color_stop = ""
-        s = super(ColorFormatter, self).format(record)
+
+    def remove_color(self, record):
         del record.color
         del record.color_stop
+
+    def format(self, record):
+        self.add_color(record)
+        s = super(ColorFormatter, self).format(record)
+        self.remove_color(record)
+        return s
+
+
+class ExtrasFormatter(logging.Formatter):
+    """Formats extra keywords into %(extras)s placeholder.
+
+    Any keywords passed to a logging call will be formatted into a
+    "extras" string and included in a logging message.
+    Example:
+        logger.info('my message', extra='keyword')
+    will cause an "extras" string of:
+        [extra: keyword]
+    to be inserted into the format in place of %(extras)s.
+
+    The optional `keywords` argument must be passed into the init
+    function to enable this functionality. Without it normal daiquiri
+    formatting will be applied. Any keywords included in the
+    `keywords` parameter will not be included in the "extras" string.
+
+    Special keywords:
+
+    keywords
+      A list of strings containing keywords to filter out of the
+      "extras" string.
+
+    extras_template
+      A format string to use instead of '[{0}: {1}]'
+
+    extras_separator
+      What string to "join" multiple "extras" with.
+
+    extras_prefix and extras_suffix
+      Strings which will be prepended and appended to the "extras"
+      string respectively. These will only be prepended if the
+      "extras" string is not empty.
+    """
+
+    def __init__(self,
+                 fmt=None,
+                 datefmt=None,
+                 style='%',
+                 keywords=None,
+                 extras_template='[{0}: {1}]',
+                 extras_separator=' ',
+                 extras_prefix=' ',
+                 extras_suffix=''):
+        self.keywords = keywords
+        self.extras_template = extras_template
+        self.extras_separator = extras_separator
+        self.extras_prefix = extras_prefix
+        self.extras_suffix = extras_suffix
+        super(ExtrasFormatter, self).__init__(fmt=fmt,
+                                              datefmt=datefmt,
+                                              style=style)
```

Comment:
It seems Python 2.7 does not like that. `logging.Formatter` does not have a `style` argument in Python 2 :(

Reply:
I've never had to support Python 2, how do I fix an issue like this?

Reply:
Lucky you.

Rewrite this like that:

```python
   def __init__(self,
                  keywords=None,
                  extras_template='[{0}: {1}]',
                  extras_separator=' ',
                  extras_prefix=' ',
                  extras_suffix='',
                  *args, **kwargs):
         [...]
         super(ExtrasFormatter, self).__init__(*args, **kwargs)
```